src/modules/tokens.ts

```diff
@@ -190,6 +190,17 @@ export namespace tokens {
 
   }
 
+  function calculateWeightedAveragePrice(prevVol: BigInt, prevAvgPrice: BigDecimal, currentVol: BigInt, currentPrice: BigDecimal): BigDecimal {
+    let prevVolumeBD = prevVol.toBigDecimal()
+    let prevValWeighted = prevVolumeBD.times(prevAvgPrice)
+    let currentVolumeBD = currentVol.toBigDecimal()
+    let currentAvgWeighted = currentVolumeBD.times(currentPrice)
+
+    let numerator = prevValWeighted.plus(currentAvgWeighted)
+    let denominator = prevVolumeBD.plus(currentVolumeBD)
+    return numerator.div(denominator) as BigDecimal
```

Comment:
Just curious, why do you need the cast to BigDecimal?
Aren't all values here BigDecimal already?

Reply:
Yeah I know that, but BigDecimal is not a native type and the compiler complains requiring an explicit type. :(